### DISCUSSION TITLE
What has changed in `response.body.cancel()` between node 20 and 21?

### DISCUSSION THREAD

**Discussion by elf-pavlik** [nodejs/undici#3208] [Q&A] [answered]
I have [an integration test for a server streaming HTTP responses](https://github.com/CommunitySolidServer/CommunitySolidServer/pull/1878/files#diff-07ded2ff42697b3d98c0b5ead2366107dddc27bc9a9a6966c3278ef36220c0ab).

This integration test worked fine on node 20 but failed on node 21 and 22.
The failure was related to the [`server.close()`](https://nodejs.org/api/http.html#serverclosecallback) not resolving due to the open connections. I'm relying on 

```js
const reader = response.body.getReader()
try {
  /* read some chunks */
}
finally {
  reader.releaseLock()
  await response.body.cancel()
}
```

At the end of the tests, `server.close()` resolves just fine on node 20.

With node 21 and 22 I had to call [`server.closeAllConnections()`](https://nodejs.org/api/http.html#servercloseallconnections) first for the `server.close()` to resolve.

In the end I changed all the tests to use

```js
const response = await fetch(iri, {
  dispatcher: new Agent({ bodyTimeout: 1000 }),
});
```
This also ensures that all the connections close and that `server.close()` resolves on nodes 21 and 22.

While this might be enough for my tests, I still would like to understand why the connections appear to stay open after `response.body.cancel()` resolves.

**Comment by mcollina** [comment]
Could you prepare a complete example to reproduce? I have a feeling on why but I would need to investigate.

**Reply by elf-pavlik** [reply]
First, an extra clarification: The client works as expected on node 20 and 22.1, while it seems to leave the connection open on node 21 and 22.0. The node version that the server is run with doesn't seem to matter.

The difference is when the scripts exit and when they don't.

server.mjs
```js
import { createServer } from 'node:http'

// create server
const server = createServer()

server.listen(5000)

server.on('request', async(req, res) => {
  res.write('hello')
  server.close()
  // if added both scripts will also exit on 21 and 22.0
  //server.closeAllConnections()
})
```

client.mjs

```js
// import { Agent } from "undici"
const url = 'http://localhost:5000'

const response = await fetch(url, {
  // if added both scripts will also exit on 21 and 22.0
  // dispatcher: new Agent({ bodyTimeout: 1000 })
})

const reader = response.body.getReader();

try {
  await reader.read()
} finally {
  reader.releaseLock();
  await response.body.cancel();
}
```

I left comments that can force the disconnect from either the server or client that also work on node 21 and 22.0

**Reply by mcollina** [reply, marked as answer]
I have a feeling we added a bug somewhat earlier that it was then fixed. I've added a test for this exact case, so we know we won't regress anymore: https://github.com/nodejs/undici/pull/3219.

You can expect the version of undici in 22.1.0 to reach 20 in the next quarter.
